Give ability for IPC calls to work with NativePromise

jyavenard opened the pull request:
#### f997a9edb80cd79caeb1d0a1b87610ffd7a56e88
<pre>
Give ability for IPC calls to work with NativePromise
<a href="https://bugs.webkit.org/show_bug.cgi?id=261720">https://bugs.webkit.org/show_bug.cgi?id=261720</a>
rdar://115704438

Reviewed by Kimmo Kinnunen.

Add the ability to send an asynchronous IPC message where the response
will be returned in the form of a NativePromise.
The NativePromise will be rejected with the IPC error if an error occurs.
Otherwise, the value returned is the value of the IPC response. Unlike the
CompletionHandler version sendWithAsyncReply, the returned value is only
wrapped in a std::tuple if more than 1 element is returned.

The Promise is defined for each asynchronous message generated like so:
```
class WaitForTarget {
public:
    using Arguments = std::tuple&lt;WebCore::SeekTarget&gt;;

    static IPC::MessageName name() { return IPC::MessageName::MediaSourcePrivateRemote_WaitForTarget; }
    static constexpr bool isSync = false;
    static constexpr bool canDispatchOutOfOrder = false;
    static constexpr bool replyCanDispatchOutOfOrder = false;

    static IPC::MessageName asyncMessageReplyName() { return IPC::MessageName::MediaSourcePrivateRemote_WaitForTargetReply; }
    static constexpr auto callbackThread = WTF::CompletionHandlerCallThread::ConstructionThread;
    using ReplyArguments = std::tuple&lt;MediaTime&gt;;
    using Promise = WTF::NativePromise&lt;MediaTime, IPC::Error, true&gt;;
    explicit WaitForTarget(const WebCore::SeekTarget&amp; target)
        : m_arguments(target)
    {
    }

    auto&amp;&amp; arguments()
    {
        return WTFMove(m_arguments);
    }

private:
    std::tuple&lt;const WebCore::SeekTarget&amp;&gt; m_arguments;
};
```

We need to introduce the concept of NativePromise::runSynchronouslyOnTarget
in order to ensure that processing of asynchronous results is kept in order
when mixing with the sendWithAsyncReply/CompletionHandler API.
Consider the following:
```
    m_connectionToWebProcess-&gt;connection().sendWithPromisedReply(Messages::Message1, m_identifier)
    -&gt;whenSettled(RunLoop::main(), __func__, WTFMove(completionHandler1));
    m_connectionToWebProcess-&gt;connection().sendWithAsyncReply(Messages::Message1, WTFMove(completionHandler2), m_identifier);
```

In this example, it is expected that completionHandler1 is run before
completionHandler2 (in the same order the IPC results were received)
By default a NativePromise will always dispatch a task to run the resolve/reject
callback, and so in the example above completionHandler1 would be run after
completionHandler2 as sendWithAsyncReply immediately calls the completionHandler.

By setting the NativePromise::Producer in RunSynchronouslyOnTarget mode,
it will immediately run the callback if already on the target thread.
Preserving the order of operations.
For the time being, RunSynchronouslyOnTarget is a response only to the
problem described above.
In the future we could imagine that it is up to the NativePromise consumer
to decide how the callbacks is to be run. But this is another problem for
another day.

API tests added for both IPC and NativePromise.

* Source/WTF/wtf/NativePromise.h:
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithPromisedReply):
(IPC::Connection::makeAsyncReplyHandler):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
(forward_declarations_and_headers):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
* Tools/TestWebKitAPI/Tests/IPC/MessageSenderTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268240@main">https://commits.webkit.org/268240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/577ca3e1d9d562819462f632f78ba5e6748c7488

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19116 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20982 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17866 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19316 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22751 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19597 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16627 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21863 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19289 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16620 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17412 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23789 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16589 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17668 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17586 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21739 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18441 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18153 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22499 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17255 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5456 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4551 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21609 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23749 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17983 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5320 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -78,8 +79,9 @@ enum class SendOption : uint8_t {
     DispatchMessageEvenWhenWaitingForSyncReply = 1 << 0,
     DispatchMessageEvenWhenWaitingForUnboundedSyncReply = 1 << 1,
     IgnoreFullySynchronousMode = 1 << 2,
+    NoDispatchToMainOnError = 1 << 3,
```
I'm not fan of introducing yet-another if arm. It'd be great if this discrepancy would be removed altogether and the semantics were well-defined from the start.
However, it's not very productive to require this for this patch..

jyavenard replied:
As discussed offline, I've removed this option and let the callback be dispatched through the main runloop on error
I also added two tests that checks that the CB is still called asynchronously as it's the intended behaviour.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -704,6 +714,20 @@ Connection::AsyncReplyID Connection::sendWithAsyncReply(T&& message, C&& complet
 }
 
+template<typename T>
+Ref<typename T::Promise> Connection::sendWithPromisedReply(T&& message, uint64_t destinationID, OptionSet<SendOption> sendOptions)
```
If I read correctly, there is a discrepancy in how the code is written currently.
The idea is that the sendWithAsyncReply callback is:
* invoked in the function dispatcher of the connection
* If there's an error, the reply handler is dispatched in the main thread

The way you add this, the implication is that:
* you construct a promise probably bound to the current function dispatcher
* However, on success the promise is resolved in the function dispatcher of the connection
* If there's an error, the promise is resolved in the current function dispatcher

So only way this is going to be as-specified by the contract is if the sendWithPromisedReply is called on the function dispatcher of the connection. Currently there's no such requirement for the sendWithAsyncReply

jyavenard replied:
A promise itself isn't bound on any threads, it can be resolved anywhere: it's thread-safe.

It's the consumer of that promise that defines on which thread it requires the callback to run.

> If there's an error, the promise is resolved in the current function dispatcher
I have fixed this so that similar to sendWithAsyncReply, "If there's an error, the reply handler is dispatched in the main thread" , from there it will immediately run if the consumer's target queue is the main thread, or dispatched to the requested thread.

I guess it would be nice to avoid the dispatch if the consumer's target queue wasn't the main thread, but Connect has no way to know what the ultimate target queue is going to be.

kkinnunen-apple reviewed (approved):
In general there's a semantic difference with using the cb API vs using the promise API, and that makes everything very complicated to reason about.
The underlying issue is of course that the CB-based API doesn't have sensible semantics.

You need to re-generate the test results for the generator, so the changes you make to messages.py are reviewable:
``` 
make -C Source/WebKit/Scripts/webkit/tests
```

jyavenard committed:
Give ability for IPC calls to work with NativePromise
https://bugs.webkit.org/show_bug.cgi?id=261720
rdar://115704438

Reviewed by Kimmo Kinnunen.

Add the ability to send an asynchronous IPC message where the response
will be returned in the form of a NativePromise.
The NativePromise will be rejected with the IPC error if an error occurs.
Otherwise, the value returned is the value of the IPC response. Unlike the
CompletionHandler version sendWithAsyncReply, the returned value is only
wrapped in a std::tuple if more than 1 element is returned.

The Promise is defined for each asynchronous message generated like so:
```
class WaitForTarget {
public:
    using Arguments = std::tuple<WebCore::SeekTarget>;

    static IPC::MessageName name() { return IPC::MessageName::MediaSourcePrivateRemote_WaitForTarget; }
    static constexpr bool isSync = false;
    static constexpr bool canDispatchOutOfOrder = false;
    static constexpr bool replyCanDispatchOutOfOrder = false;

    static IPC::MessageName asyncMessageReplyName() { return IPC::MessageName::MediaSourcePrivateRemote_WaitForTargetReply; }
    static constexpr auto callbackThread = WTF::CompletionHandlerCallThread::ConstructionThread;
    using ReplyArguments = std::tuple<MediaTime>;
    using Promise = WTF::NativePromise<MediaTime, IPC::Error, true>;
    explicit WaitForTarget(const WebCore::SeekTarget& target)
        : m_arguments(target)
    {
    }

    auto&& arguments()
    {
        return WTFMove(m_arguments);
    }

private:
    std::tuple<const WebCore::SeekTarget&> m_arguments;
};
```

We need to introduce the concept of NativePromise::runSynchronouslyOnTarget
in order to ensure that processing of asynchronous results is kept in order
when mixing with the sendWithAsyncReply/CompletionHandler API.
Consider the following:
```
    m_connectionToWebProcess->connection().sendWithPromisedReply(Messages::Message1, m_identifier)
    ->whenSettled(RunLoop::main(), __func__, WTFMove(completionHandler1));
    m_connectionToWebProcess->connection().sendWithAsyncReply(Messages::Message1, WTFMove(completionHandler2), m_identifier);
```

In this example, it is expected that completionHandler1 is run before
completionHandler2 (in the same order the IPC results were received)
By default a NativePromise will always dispatch a task to run the resolve/reject
callback, and so in the example above completionHandler1 would be run after
completionHandler2 as sendWithAsyncReply immediately calls the completionHandler.

By setting the NativePromise::Producer in RunSynchronouslyOnTarget mode,
it will immediately run the callback if already on the target thread.
Preserving the order of operations.
For the time being, RunSynchronouslyOnTarget is a response only to the
problem described above.
In the future we could imagine that it is up to the NativePromise consumer
to decide how the callbacks is to be run. But this is another problem for
another day.

API tests added for both IPC and NativePromise.

* Source/WTF/wtf/NativePromise.h:
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithPromisedReply):
(IPC::Connection::makeAsyncReplyHandler):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
(forward_declarations_and_headers):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
* Tools/TestWebKitAPI/Tests/IPC/MessageSenderTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268240@main
